internal/provider/resource_gitlab_project_protected_environment.go

```diff
@@ -0,0 +1,198 @@
+package provider
+
+import (
+	"context"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	gitlab "github.com/xanzy/go-gitlab"
+)
+
+// https://docs.gitlab.com/ee/ci/environments/protected_environments.html
+var _ = registerResource("gitlab_project_protected_environment", func() *schema.Resource {
+	// lintignore: XR002 // TODO: Resolve this tfproviderlint issue
+	return &schema.Resource{
+		Description: "This resource allows you to create and manage Project Access Token for your GitLab projects.",
+
+		CreateContext: resourceGitlabProjectProtectedEnvironmentCreate,
+		ReadContext:   resourceGitlabProjectProtectedEnvironmentRead,
+		DeleteContext: resourceGitlabProjectProtectedEnvironmentDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Description:  "The ID or full path of the project which the protected environment is created against.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"environment": {
+				Description:  "The name of the environment.",
+				Type:         schema.TypeString,
+				ForceNew:     true,
+				Required:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"deploy_access_levels": {
+				Description: "Array of access levels allowed to deploy, with each described by a hash.",
+				Type:        schema.TypeList,
+				ForceNew:    true,
+				Required:    true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"access_level": {
+							Description:  "Level of access required to deploy to this protected environment (developer, maintainer)",
```

Comment:
Use the render function as here:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/32e9fe228f6d0a3b8feafe407f2286b36a55f927/internal/provider/resource_gitlab_project_membership.go#L47

Reply:
... consistent with the validation